nemo/core/optim/distributed_adam.py

```diff
@@ -318,3 +325,24 @@ def load_state_dict(self, state_dict):
                 old_main_param.copy_(new_main_param.detach())
             del state_dict['fp32_optim_fp32_params']
         return super().load_state_dict(state_dict)
+
+    def sharded_state_dict(self, model_sharded_state_dict):
+        optimizer_state_dict = self.state_dict()
+
+        id_to_sharded_param_map = get_param_id_to_sharded_param_map(
+            model_sharded_state_dict=model_sharded_state_dict,
+            optim_params_iter=self.parameters(with_fp32_optim_params=True),
+        )
+        # Convert state
+        step = optimizer_state_dict['state'].pop('step')
+        optim_state_to_sharding_state(optimizer_state_dict, id_to_sharded_param_map)
+        optimizer_state_dict['state']['step'] = step
+
+        def rename_fp32_params(x):
+            if isinstance(x, ShardedTensor) and x.key.startswith('optimizer.state.param'):
+                x.key = x.key.replace('optimizer.state.param', 'optimizer.state.fp32_from_fp16')
```

Comment:
@ericharper we could choose a different prefix here, as long as we change it [here](https://github.com/NVIDIA/NeMo/pull/7281/files#diff-7667eae242a8ef776bff78cd08e79bc81df4896a450f0a781f6ed317a3dfb7ffR254) as well